Add configure endpoints documentation (#967)
wso2/docs-choreo-dev: en/docs/develop-components/configure-endpoints-body.md

Services and integrations are exposed to other services, integrations, or applications through endpoints. A service or an integration can expose multiple endpoints, each representing a unique entry point into the service. For example, a service may expose a REST API endpoint and a GraphQL endpoint, each providing different ways to interact with the service. Endpoints provide specific details for how the service or the integration can be consumed. For instance, the port number, protocol, and the schema such as open API specification (OAS) or GraphQL schema. By defining these details, endpoints make it possible for other services, integrations, and applications to discover and interact with the service in a standardized way.

> Choreo defines endpoints by combining port binding, protocol, endpoint name, network visibility, endpoint schema, and additional protocol-related fields. Following table describe each attribute of an endpoint.

Reviewer: ```suggestion
Choreo defines endpoints by combining port binding, protocol, endpoint name, network visibility, endpoint schema, and additional protocol-related fields. The following table describes each attribute of an endpoint.
```

| Field | Description |
| ----- | ----------- |
| Name | A unique identifier for the endpoint within the service component. |
| Port | The network port on which the endpoint is accessible. |
| Type | The communication protocol used by the endpoint (e.g., HTTP, HTTPS, gRPC, etc.). |
| Network Visibility | Determines the level of visibility of an endpoint. Possible values are: <ul><li>Project: Allows components within the same project to access the endpoint.</li><li>Organization: Allows any component within the same organization to access the endpoint but restricts access to components outside the organization.</li><li>Public: Allows any client to access the endpoint, regardless of location or organization.</li></ul> |
| Schema | specifies the structure and format of the data exchanged through the endpoint. |
| Context (HTTP and GraphQL only) | A context path that you add to the endpoint's URL for routing purposes. |

## Configure endpoints

> The method of defining endpoints depends on the build preset. There are three build presets the supports defining endpoints: 

Reviewer: Might need to change "the" to "that" in the second sentence or rephrase it.

Author: fixed.

- Ballerina
- Dockerfile
- WSO2 MI

### Configure endpoints with the Dockerfile build-preset

When you build a service component using the Dockerfile build-preset, you can configure the endpoint details with the endpoints.yaml configuration file. You must place this file inside the `.choreo` directory at the build context path and commit it to the source repository.

> See [Understanding the endpoints.yaml file](#understanding-the-endpointsyaml-file) to learn about the endpoints.yaml file.

Reviewer: Shall we have `endpoints.yaml` in code format? Might have to apply this in all applicable places.

Author: done

### Configure endpoints with the Ballerina build-preset

When you create a service component with the `Ballerina preset`, Choreo automatically detects the endpoint details for REST API and GraphQL endpoints. You can override the auto-generated endpoint configuration by providing the endpoints.yaml file in the source directory

See [Understanding the endpoints.yaml file](#understanding-the-endpointsyaml-file) to learn about the endpoints.yaml file.

### Configure endpoints with the WSO2 MI build-preset

> WSO2 MI build preset is where you can deploy Integrations developed with WSO2 Micro Integrator as an API. In this preset, you have three different ways to define endpoints. Choreo gives priory to the definition of endpoints in the below-mentioned order. 

Reviewer: Looks like the i in integrations is mistakenly capitalized.

Author: fixed

3. **Provide default endpoints**
If endpoints.yaml is not provided and if the source Micro Integrator project doesn't have APIs, Choreo generates a default endpoint which will expose the default micro integrator port (8290) with `Project` visibility and wildcard context.

### Learn the endpoints.yaml file

The endpoints.yaml file has a specific structure and contains the following details:

| Field                | Required     | Description                                                                      |
|----------------------|--------------|----------------------------------------------------------------------------------|
| **version**          | Required     | The version of the endpoints.yaml file.                                           |
| **name**             | Required     | A unique name for the endpoint, which Choreo will use to generate the managed API.|
| **port**             | Required     | The numeric port value that gets exposed via this endpoint.                      |

> | **type**             | Required     | The type of traffic this endpoint is accepting, such as `REST`, `GraphQL`, or `gRPC`.Currently the MI preset supports only the `REST` type.                                         |

Reviewer: ```suggestion
| **type**             | Required     | The type of traffic this endpoint is accepting, such as `REST`, `GraphQL`, or `gRPC`. Currently, the MI preset supports only the `REST` type.                                         |
```

Author: fixed

| **networkVisibility**| Required     | The network level visibility of this endpoint, which defaults to `Project` if not specified. Accepted values are `Project`, `Organization`, or `Public`.|
| **context**          | Required     | The context (base path) of the API that Choreo exposes via this endpoint.        |

> | **schemaFilePath**   | Required     | The swagger definition file path. Defaults to the wildcard route if not provided. this field should be a relative path to the project path when using the MI build preset. For REST endpoint types when using the Ballerina or Dockerfile preset, this field should be a relative path to the component root or Docker context .|

Reviewer: ```suggestion
| **schemaFilePath**   | Required     | The swagger definition file path. Defaults to the wildcard route if not provided. This field should be a relative path to the project path when using the MI build preset. For REST endpoint types when using the Ballerina or Dockerfile preset, this field should be a relative path to the component root or Docker context.|
```

Author: fixed

#### Sample endpoints.yaml

**File location**:

```bash
<docker-build-context-path>/.choreo/endpoints.yaml
```

!!! note
    - For components built with Ballerina build preset `docker-build-context-path` should be replaced with `component-root`. 
    For example: `<component-root>/.choreo/endpoints.yaml`

    - For components built with WSO2 MI build preset `docker-build-context-path` should be replaced with `<Project Path>`. 
    For example: `<Project Path>/.choreo/endpoints.yaml`

> **File Content**:

Reviewer: ```suggestion
**File content**:
```

Author: fixed

```yaml
# +required Version of the endpoint configuration YAML
version: 0.1

# +required List of endpoints to create
endpoints:
  # +required Unique name for the endpoint. (This name will be used when generating the managed API)
- name: Greeting 9090
  # +required Numeric port value that gets exposed via this endpoint
  port: 9090
  # +required Type of the traffic this endpoint is accepting. Example: REST, GraphQL, etc.
  # Allowed values: REST, GraphQL, GRPC
  type: REST
  # +optional Network level visibility of this endpoint. Defaults to Project
  # Accepted values: Project|Organization|Public.
  networkVisibility: Project
  # +optional Context (base path) of the API that is exposed via this endpoint.
  # This is mandatory if the endpoint type is set to REST or GraphQL.
  context: /greeting
  # +optional Path to the schema definition file. Defaults to wild card route if not provided
  # This is only applicable to REST endpoint types.
  # The path should be relative to the docker context.
  schemaFilePath: greeting_openapi.yaml
```

> ## Expose endpoint as managed APIs

Reviewer: Should be either 
Expose endpoints as managed APIs
or
Expose an endpoint as a managed API

Author: fixed

> Exposing endpoints as managed APIs is crucial to ensure secure and controlled access to the services being exposed. When a user wants to expose their written service to the outside world or to the organization at large, there is an inherent security risk involved. To mitigate this risk, Choreo platform is build with an internal (access within organization only) or external (publicly accessible) gateway that is protected with Choreo API management making the services secure by design.

Reviewer: ```suggestion
Exposing endpoints as managed APIs is crucial to ensure secure and controlled access to the services being exposed. When a user wants to expose their written service to the outside world or to the organization at large, there is an inherent security risk involved. To mitigate this risk, the Choreo platform is built with an internal (access within the organization only) or external (publicly accessible) gateway that is protected with Choreo API management making the services secure by design.
```

Author: fixed
